nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py

```diff
@@ -369,35 +369,88 @@ def allreduce_gradients(self):
-                    and parallel_state.get_pipeline_model_parallel_world_size() > 1
-                    and parallel_state.get_pipeline_model_parallel_split_rank() is not None
-                    and self.cfg.get('position_embedding_type') == 'learned_absolute'
-                ):
-                    position_embeddings_weight = self.enc_dec_model.position_embeddings_weight()
+        # All-reduce position embeddings for T5.
+        if (
+            parallel_state.is_rank_in_position_embedding_group()
+            and parallel_state.get_pipeline_model_parallel_world_size() > 1
+            and parallel_state.get_pipeline_model_parallel_split_rank() is not None
+            and self.cfg.encoder.get('position_embedding_type') == 'learned_absolute'
+            and self.cfg.decoder.get('position_embedding_type') == 'learned_absolute'
+        ):
+            if self.enc_dec_model.share_token_embeddings:
+                position_embeddings_weight = self.enc_dec_model.position_embeddings_weight()
+                if self.megatron_amp_o2:
+                    grad = position_embeddings_weight.main_grad
+                else:
+                    grad = position_embeddings_weight.grad
+                torch.distributed.all_reduce(grad, group=parallel_state.get_position_embedding_group())
+
+        # All-reduce relative position embeddings for T5.
+        if (
+            parallel_state.get_pipeline_model_parallel_world_size()
+            > 2  # This > 2 and not > 1 since with PP=2 encoder RPE can live only on one rank.
+            and parallel_state.get_pipeline_model_parallel_split_rank() is not None
+        ):
+            # For split rank = 1, we have only one encoder rank and so we don't need to allreduce.
+            if (
+                self.cfg.encoder.get('position_embedding_type') == 'relative'
+                and parallel_state.is_rank_in_encoder_relative_position_embedding_group()
+                and parallel_state.get_pipeline_model_parallel_split_rank() > 1
+            ):
+                position_embeddings_weight = self.enc_dec_model.encoder_relative_position_embeddings_weight()
+                if self.megatron_amp_o2:
+                    grad = position_embeddings_weight.main_grad
+                else:
+                    grad = position_embeddings_weight.grad
+                torch.distributed.all_reduce(
+                    grad, group=parallel_state.get_encoder_relative_position_embedding_group()
+                )
+
+            # For split rank == pipeline_world_size - 1, we have only one decoder rank and so we don't need to allreduce.
+            if (
+                self.cfg.decoder.get('position_embedding_type') == 'relative'
+                and parallel_state.is_rank_in_decoder_relative_position_embedding_group()
+                and parallel_state.get_pipeline_model_parallel_split_rank()
+                < parallel_state.get_pipeline_model_parallel_world_size() - 1
```

Comment:
We decided these lines are unnecessary and reduce the code readability so will remove.